Add new LanguagePicker component

jsnajdr opened the pull request:
This is a basic `LanguagePicker` component that's been split off the big #12544 PR -- see the comments there to learn what's the big picture.

![language-picker-recording](https://cloud.githubusercontent.com/assets/664258/25040248/6a684800-2108-11e7-9dfc-e6eed5b0e920.gif)

It's not yet used anywhere in the UI, but can be tested on the [example in devdocs](https://calypso.live/devdocs/design/language-picker?branch=add/language-picker)

The language picker modal is pretty basic in this first version -- more features like suggested languages and grouping languages by territory will be coming in follow-up PRs.

a8ci18n commented on client/components/language-picker/modal.jsx:
```diff
@@ -0,0 +1,189 @@
+			},
+			{
+				action: 'confirm',
+				label: translate( 'Select Language' ),
```
Hi! I've found a possible matching string that has already been translated **60** times:
`translate( 'Select a language below' )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/language-picker)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

folletto commented:
Design wise: 👍

rickybanister commented:
@folletto—It's still not obvious to me that you can scroll that list—especially when the first cut off row is completely off screen. Can we try something like our white blur fade out effect across the bottom of the box and ensure that some amount of that next row of options is visible?

folletto commented:
Sure. Let's do that in a separate PR 👍

blowery commented:
We've been kicking around adding a lang picker to the search results for Reader (and possibly tag streams). This looks neat, but I think we'd need a different, more compact UI treatment. Do you this would be suitable to reuse or adapt for that purpose?

/cc @fraying

folletto commented:
> This looks neat, but I think we'd need a different, more compact UI treatment. Do you this would be suitable to reuse or adapt for that purpose?

I guess depends... What would be the need?

jsnajdr commented:
Pushed a new version with some CSS improvements:
- rewrote the modal layout to `flex` with `column` direction, removed the previous styling with `position: absolute` and exact pixel sizes. Tested on all current browsers including IE11 and it works :+1:
- tuned the styling of the language list items and the selected state

yoavf commented:
Edge case, but the language name for Hebrew is spelled with diacritics "עִבְרִית"
Searching for just "עברית" (no diacritics, as would everyone actually do) returns no result. 
I'm ok with changing the language name at the source in this case, but a safer change would maybe  be to have the search ignore some utf-8 ranges of characters.

I nice feature would be for the search to also retrieve the language based on the English name. So searching for "Hebrew" would return "עִבְרִית"

folletto commented:
> rewrote the modal layout to flex

Excellent idea! 🎉 

> I nice feature would be for the search to also retrieve the language based on the English name. So searching for "Hebrew" would return "עִבְרִית"

That's the plan! ;)
See #12544 and the related design for details.

jsnajdr commented:
> I nice feature would be for the search to also retrieve the language based on the English name. So searching for "Hebrew" would return "עִבְרִית"

That's a planned feature -- the CLDR database has all the data needed for that.

> Edge case, but the language name for Hebrew is spelled with diacritics "עִבְרִית"
Searching for just "עברית" (no diacritics, as would everyone actually do) returns no result.
I'm ok with changing the language name at the source in this case, but a safer change would maybe be to have the search ignore some utf-8 ranges of characters.

This is an excellent idea - in Czech, one can also search for "cestina" and expect to find "čeština". 

But the implementation could be challenging in a JavaScript environment :) In Java, which has perfect Unicode support in the standard library, I can decompose the accented characters into letter+accent pairs with `java.text.Normalizer`, and then filter out the accents with a special character class in a regexp. I'm afraid this won't be so easy in JS without including some giant 3rd party Unicode library.

blowery commented:
> I guess depends... What would be the need?

At the moment, tag streams and search results are restricted to the user's chosen language. For folks that read more than one language, it would be nice to be able to show results in a different language. For example, Sirin has her user lang set to EN but would like to see posts in German, without going through the hassle of altering her user lang. 

It would appear somewhere on https://wordpress.com/read/search?q=hotcakes and https://wordpress.com/tag/hotcakes 

The current UI for the selector is pretty large, so I'm not sure where it would fit in on those pages.

blowery commented:
FWIW, I don't think my concerns should hold up the PR in any way. We can always address it down the road.

folletto commented:
> The current UI for the selector is pretty large, so I'm not sure where it would fit in on those pages.

If the UI can be triggered as a modal like this one, we can design a variation that uses the same pattern but opens a UI with checkboxes instead of single selections.

But yes, let's discuss/design this separately. :)

deBhal commented on client/components/language-picker/README.md:
```diff
@@ -0,0 +1,69 @@
+`valueKey` - **optional** Which property of the `languages` objects should be used
+as the `value` of the component? The default is `value`, other options are `langSlug` and
+`wpLocale`. This determines what kind of value is expeced as the `value` property and
+what is the value passed as parameter in the `onChange` handler.
```
excpeced -> expected.

We normally describe what the prop is or does ( e.g. "Determines which property of the `languages` objects should be used…"

You could say it more succinctly like "Should be one of `value` (default), `langSlug` or `wpLocale`". (as a bonus, it also suggests the [propType syntax](https://github.com/reactjs/prop-types/blob/b06161d16f553bfbf006fc3d23f22531771efe3e/factoryWithTypeCheckers.js#L66))

More generally, this is pretty confusing. What's the backstory here? Is "kind of value" the type? There's a comment in the code about expecting numbers sometimes, is this about that?

jsnajdr replied:
The backstory is: every record in the `languages` list has this structure:
```
{
	name: 'English',
	value: 1,
	langSlug: 'en',
	wpLocale: 'en_US'
}
```
There's `name`, the string displayed in UI, and three unique identifiers of the language. The numerical `value` is used when setting a site language, the ISO-compliant identifier `langSlug` is used in user settings, and there is also `wpLocale`.

The `valueKey` prop determines which of these identifiers is used as the `value` prop, and what is fired in the `onChange` callback.

An alternative way how to implement this would be to transform the `languages` list before passing it to the component:
```
const actualLanguages = languages.map( lang => ( {
  name: lang.name,
  value: lang[valueKey]
} );
```
I'll try to document the `valueKey` better.

folletto commented:
> it feels a bit weird on a large screen

> would it make sense to have a "big" case where we have a single tab with the popular languages above all the languages?

Think in terms of goals, not in terms of "what could fit": the reason we show a limited range to start with is to help people choose by avoiding overwhelming them with too many meaningless options. Having more space doesn't change that, so we shouldn't cram all the languages in.

The solution here is to limit height and width of the modal beyond a certain size. We can stop the modal to grow at some point, maybe around ~1200px or so, and attempting keeping the height proportional to that (any approach that approximates ~4:3 maybe).


> It would also be nice to put the keyboard focus onto the search bar when I open the modal so I can just start typing. There's nowhere else for my keystrokes to go. Combined with the above, that offers a nice keyboard driven UX.

This is a good advice, but I wouldn't pre-focus as it would open the search bar and hide the filters. Ideally we should just trap any character (i.e. not special keys) keyboard interaction and if we detect it, we open search. 

—

Given the scope of this, and the current status of the component, I'd keep the two above as separate PRs / feature requests to avoid slowing down the merge of the component.

deBhal commented:
Thanks @folletto, that's a great response. I agree on all points.

jsnajdr commented on client/components/language-picker/modal.jsx:
```diff
@@ -0,0 +1,189 @@
+	renderTabItems() {
+		const tabs = [ 'popular', '' ];
+
+		return map( tabs, filter => {
```
The `getFilterLabel` method is also used when rendering the `<SectionNav>` `selectedText` attribute. So it's not so easy to drop it.

There will be more tabs in this list when we implement the grouping of languages by territory (Europe, Africa, ...) We can do some refactoring at that time.

deBhal replied:
> We can do some refactoring at that time.

That'll work 👍 .

jsnajdr commented on client/components/language-picker/README.md:
```diff
@@ -0,0 +1,69 @@
+			/>
+		);
+	}
+}
```
The example code in README now contains connecting to the `state/user-settings` Redux store :)

deBhal replied:
Awesome!

jsnajdr commented:
Thanks @deBhal for the review! I tried to rewrite the `valueKey` documentation, let me know if it makes more sense now.

Also, I added a Redux usage to the code sample in README. The `user-settings` Redux state wasn't finished yet when I was writing that README :)

You have other great suggestions that would be great to implement as separate PRs -- start searching on first keystroke, tune the size of the modal on large screens... They could be useful not only for the `LanguagePicker` modal itself, but also for the Media and Contact Form modals in the Editor.

deBhal commented on client/components/language-picker/README.md:
````diff
@@ -0,0 +1,75 @@
+
+## Example Usage
+
+```js
````
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

deBhal commented on client/components/language-picker/README.md:
```diff
@@ -0,0 +1,75 @@
+	}
+];
+
+class LanguagePickerDemo {
```
needs `extends PureComponent`

deBhal commented:
> The backstory is: every record in the `languages` list has this structure:
> ```
> {
> 	name: 'English',
> 	value: 1,
> 	langSlug: 'en',
> 	wpLocale: 'en_US'
> }
> ```
> There's `name`, the string displayed in UI, and three unique identifiers of the language. The > numerical `value` is used when setting a site language, the ISO-compliant identifier `langSlug` is > used in user settings, and there is also `wpLocale`.
> 
> The `valueKey` prop determines which of these identifiers is used as the `value` prop, and what is fired in the `onChange` callback.
> 
> An alternative way how to implement this would be to transform the `languages` list before passing it to the component:
> ```
> const actualLanguages = languages.map( lang => ( {
>   name: lang.name,
>   value: lang[valueKey]
> } );
> ```
> I'll try to document the `valueKey` better.

Thanks for the explanation, that's much clearer now. The `valueKey` is about maintaining some distance between `config.languages` and the picker itself.

For the purposes of this PR, it works well.

> Thanks @deBhal for the review! I tried to rewrite the valueKey documentation, let me know if it makes more sense now.

👍 

> Also, I added a Redux usage to the code sample in README. The user-settings Redux state wasn't finished yet when I was writing that README :)

Heh, that works out well :) Looks good!

> You have other great suggestions that would be great to implement as separate PRs -- start searching on first keystroke, tune the size of the modal on large screens...
> They could be useful not only for the LanguagePicker modal itself, but also for the Media and Contact Form modals in the Editor.

Definitely. Sorry for the feature creep :p Doing the behaviours in a reusable way is a cool thought.

I stuck the example code into the devdocs, and it's missing a couple of things to run. Once those are in place, though, it works great. Looks like the redux updates means you've already hit the dynamic update bonus objective 🎊 :

![languagepicker](https://cloud.githubusercontent.com/assets/5952255/25319739/54cdb06e-28e4-11e7-9401-3f1be854b6d5.gif)

Musing about the future, the component is not far off being a completely generic tabbed picker. If I passed in `{ 'fruit: apple, name: 'Apple'}, { fruit: 'banana', name: 'Banana' }` and set the `valueKey` to `'fruit'`, it would work just fine, and I could respond to the user's choice of fruit, right?

We'd need to expose the last few strings as props and rework the tabs a little (predicates for filters, I think), and add props for some strings like the buttons and placeholder.

I'd also think about using a `{ key: { name: }, ... }` format instead of `valueKey`. That might simplify the component and emphasize the value a little more. It's easy to get there from `config.languages` with with [lodash.keyBy](https://lodash.com/docs/4.17.4#keyBy) and has some nice side-effects, like having a unique key to give to react child components.

However, none of that needs to go in this PR.

Please add the missing `import`s and `extends`, and then go ahead and ship it :ship:. With the changes being documentations only, I'll add the "Ready for Merge" now to save wasting any more time.

Good work!

jsnajdr commented:
Thanks @deBhal for the review! I pushed a documentation update with the missing `import`s and `extends` and I'll merge the PR.

I love the idea with the generic tabbed picker. I'll definitely move in that direction when I'm adding more features to the picker - like suggested languages, more tabs with territories, advanced search... The generic picker's API will need to be able to accomodate all these features.
